README
Zita-alsa-pcmi provides easy access to ALSA PCM devices, taking
care of the many operations required to open, initialise and use
an ALSA hw: device in memory mapped mode. To applications PCM
data appears as single precision floating point samples in the
range -1.0 to 1.0f.


Release 0.3.2  (25/08/2018)
---------------------------

Maintenance release.


Release 0.1.1  (26/01/2012)
---------------------------

* This release is almost API compatible with clalsadrv-2.x.x.
  The only changes your source code will need are:

  - Change the include file.
  - Change the type of any interface objects.
  - Replace calls to stat() by state(). 
  - If you want error reporting on stderr, add an optional
    parameter to the constructor. See include file for details.

* Added support for big-endian PCM formats.

* Added support for reading and writing interleaved user buffers.

* Error messages on stderr can be selectively enabled. If an app
  is compiled without them, they can be re-enabled at runtime by
  defining the environment variable ZITA_ALSA_PCMI_DEBUG, so they
  are now by default off. See source code for details.

* Two simple demo programs are provided, one of them is the ALSA
  version of jack_delay. Complete documentation will follow later.

-- 
FA


debian/changelog
zita-alsa-pcmi (0.3.2-1.1) unstable; urgency=medium

  [ Helmut Grohne ]
  * Fix FTCBFS: Let dh_auto_build pass cross tools to make.
    (Closes: #950750)

  [ Dennis Braun ]
  * Add me as uploader

 -- Dennis Braun <d_braun@kabelmail.de>  Mon, 10 Feb 2020 18:45:22 +0100

zita-alsa-pcmi (0.3.2-1) unstable; urgency=medium

  * Team upload

  [ Felipe Sateler ]
  * Change maintainer address to debian-multimedia@lists.debian.org

  [ Ondřej Nový ]
  * Use debhelper-compat instead of debian/compat
  * d/control: Set Vcs-* to salsa.debian.org
  * d/copyright: Use https protocol in Format field

  [ Dennis Braun ]
  * New upstream version 0.3.2
  * d/control:
    + Bump dh-compat to 12
    + Bump Standards-Version to 4.5.0
    + Set RRR: no
    + Use https protocol for homepage
  * d/copyright: Update year, http > https & add myself
  * d/patches: Update patchset
  * d/rules: Replace libs/ with source/
  * d/source/local-options: Drop, obsolete
  * d/watch: Use https protocol
  * Fix zita-alsa-pcmi-utils.install

 -- Dennis Braun <d_braun@kabelmail.de>  Mon, 03 Feb 2020 20:36:38 +0100

zita-alsa-pcmi (0.2.0-4) unstable; urgency=medium

  * Set dh/compat 10.
  * Update copyright file.
  * Sign tags.
  * Fix VCS fields.
  * Bump Standards.
  * Fix hardening.
  * Avoid useless linking.
  * Remove unused substitution variable.
  * Add spelling patch.

 -- Jaromír Mikeš <mira.mikes@seznam.cz>  Tue, 27 Dec 2016 12:03:32 +0100

zita-alsa-pcmi (0.2.0-3) unstable; urgency=medium

  * Team upload.
  * Remove myself from the Uploaders field.

 -- Alessio Treglia <alessio@debian.org>  Tue, 01 Nov 2016 18:44:57 +0000

zita-alsa-pcmi (0.2.0-2) unstable; urgency=low

  [ Alessio Treglia ]
  * Add symbols file, don't use static shlibs anymore.

  [ IOhannes m zmoelnig ]
  * Adding manpages (Closes: #716677)
  * Fix usage-message of sample-applications.
  * Use canonical Vcs-* stanzas.

  [ Jaromír Mikeš ]
  * Removed DMUA.
  * Bump Standards.
  * Added local-options file.

 -- Jaromír Mikeš <mira.mikes@seznam.cz>  Tue, 06 Aug 2013 10:56:12 +0200

zita-alsa-pcmi (0.2.0-1) unstable; urgency=low

  * New upstream release.
  * Refresh patches.
  * Update debian/copyright.
  * Bump debhelper to 9.
  * Bump Standards.

 -- Alessio Treglia <alessio@debian.org>  Fri, 23 Mar 2012 12:40:59 +0100

zita-alsa-pcmi (0.1.2-1) unstable; urgency=low

  [ Jaromír Mikeš ]
  * New upstream release

  [ Alessio Treglia ]
  * Add DM-Upload-Allowed: yes.
  * This release should fix FTBFS on most architectures.

 -- Jaromír Mikeš <mira.mikes@seznam.cz>  Mon, 06 Feb 2012 14:34:50 +0100

zita-alsa-pcmi (0.1.1-1) unstable; urgency=low

  * Initial release (Closes: #657861)

 -- Jaromír Mikeš <mira.mikes@seznam.cz>  Sun, 29 Jan 2012 17:29:03 +0100


apps/mtdm.h
// ----------------------------------------------------------------------------
//
//  Copyright (C) 2006-2018 Fons Adriaensen <fons@linuxaudio.org>
//    
//  This program is free software; you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation; either version 3 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
// ----------------------------------------------------------------------------


#ifndef __MTDM_H
#define __MTDM_H


#include <unistd.h>


class Freq
{
public:

    int   p;
    int   f;
    float xa;
    float ya;
    float x1;
    float y1;
    float x2;
    float y2;
};


class MTDM
{
public:

    MTDM (int fsamp);
    int process (size_t len, float *inp, float *out);
    int resolve (void);
    void invert (void) { _inv ^= 1; }
    int    inv (void) { return _inv; }
    double del (void) { return _del; }
    double err (void) { return _err; }

private:

    double  _del;
    double  _err;
    float   _wlp;
    int     _cnt;
    int     _inv;
    Freq    _freq [13];
};


#endif



apps/mtdm.cc
// ----------------------------------------------------------------------------
//
//  Copyright (C) 2006-2018 Fons Adriaensen <fons@linuxaudio.org>
//    
//  This program is free software; you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation; either version 3 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
// ----------------------------------------------------------------------------


#include <math.h>
#include "mtdm.h"


MTDM::MTDM (int fsamp) : _cnt (0), _inv (0)
{
    int   i;
    Freq  *F;

    _freq [0].f  = 4096;
    _freq [1].f  = 2048;
    _freq [2].f  = 3072;
    _freq [3].f  = 2560;
    _freq [4].f  = 2304;
    _freq [5].f  = 2176; 
    _freq [6].f  = 1088;
    _freq [7].f  = 1312;
    _freq [8].f  = 1552;
    _freq [9].f  = 1800;
    _freq [10].f = 3332;
    _freq [11].f = 3586;
    _freq [12].f = 3841;
    _wlp = 200.0f / fsamp;
    for (i = 0, F = _freq; i < 13; i++, F++)
    {
	F->p = 128;
	F->xa = F->ya = 0.0f;
	F->x1 = F->y1 = 0.0f;
	F->x2 = F->y2 = 0.0f;
    }
}


int MTDM::process (size_t len, float *ip, float *op)
{
    int    i;
    float  vip, vop, a, c, s;
    Freq   *F;

    while (len--)
    {
        vop = 0.0f;
	vip = *ip++;
	for (i = 0, F = _freq; i < 13; i++, F++)
	{
	    a = 2 * (float) M_PI * (F->p & 65535) / 65536.0; 
	    F->p += F->f;
	    c =  cosf (a); 
	    s = -sinf (a); 
	    vop += (i ? 0.01f : 0.20f) * s;
	    F->xa += s * vip;
	    F->ya += c * vip;
	} 
	*op++ = vop;
	if (++_cnt == 16)
	{
	    for (i = 0, F = _freq; i < 13; i++, F++)
	    {
		F->x1 += _wlp * (F->xa - F->x1 + 1e-20);
		F->y1 += _wlp * (F->ya - F->y1 + 1e-20);
		F->x2 += _wlp * (F->x1 - F->x2 + 1e-20);
		F->y2 += _wlp * (F->y1 - F->y2 + 1e-20);
		F->xa = F->ya = 0.0f;
	    }
            _cnt = 0;
	}
    }

    return 0;
}


int MTDM::resolve (void)
{
    int     i, k, m;
    double  d, e, f0, p;
    Freq    *F = _freq;

    if (hypot (F->x2, F->y2) < 0.001) return -1;
    d = atan2 (F->y2, F->x2) / (2 * M_PI);
    if (_inv) d += 0.5;
    if (d > 0.5) d -= 1.0;
    f0 = _freq [0].f;
    m = 1;
    _err = 0.0;
    for (i = 0; i < 12; i++)
    {
	F++;
	p = atan2 (F->y2, F->x2) / (2 * M_PI) - d * F->f / f0;
        if (_inv) p += 0.5;
	p -= floor (p);
	p *= 2;
	k = (int)(floor (p + 0.5));
	e = fabs (p - k);
        if (e > _err) _err = e;
        if (e > 0.4) return 1; 
	d += m * (k & 1);
	m *= 2;
    }  
    _del = 16 * d;

    return 0;
}




apps/alsa_delay.cc
// ----------------------------------------------------------------------------
//
//  Copyright (C) 2006-2018 Fons Adriaensen <fons@linuxaudio.org>
//    
//  This program is free software; you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation; either version 3 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
// ----------------------------------------------------------------------------


#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <math.h>
#include <signal.h>
#include <zita-alsa-pcmi.h>
#include "pxthread.h"
#include "mtdm.h"


class Audiothr;

static Audiothr  *audio = 0;


class Audiothr : public Pxthread
{
public:

    Audiothr (const char *playdev, const char *captdev,
	      int fsamp, int frsize, int nfrags, int inp, int out);
    virtual ~Audiothr (void);

    void stop (void) { _stop = true; }
    bool running (void) const { return !_stop; }
    MTDM *mtdm (void) const { return (MTDM *) _mtdm; }

private:

    virtual void thr_main (void);

    Alsa_pcmi  *_pcmi;
    MTDM       *_mtdm;
    bool        _stop;
    int         _frsize;
    float      *_ipbuf;
    float      *_opbuf;
    int         _inp;
    int         _out;
};


Audiothr::Audiothr (const char *playdev, const char *captdev,
		    int fsamp, int frsize, int nfrags, int inp, int out) :
    _pcmi (0),
    _mtdm (0),
    _stop (true),
    _frsize (frsize),
    _ipbuf (0),
    _opbuf (0),
    _inp (inp),
    _out (out)
{
    _mtdm  = new MTDM (fsamp);
    _ipbuf = new float [frsize];
    _opbuf = new float [frsize];

    // Create and initialise the audio device.
    _pcmi = new Alsa_pcmi (playdev, captdev, 0, fsamp, frsize, nfrags, 0);
    if (_pcmi->state ()) 
    {
	fprintf (stderr, "Can't open ALSA device\n");
        delete _pcmi;
        exit (0);
    }
    _pcmi->printinfo ();
    _stop = false;

    // Run a real-time thread to process audio.
    if (thr_start (SCHED_FIFO, -20, 0x20000))
    {
 	fprintf (stderr, "Can't run in RT mode, trying normal scheduling.\n");
        if (thr_start (SCHED_OTHER, 0, 0x20000))
        {
 	    fprintf (stderr, "Can't start audio thread.\n");
	    exit (1);
	}
    }
}


Audiothr::~Audiothr (void)
{
    delete[] _ipbuf;
    delete[] _opbuf;
    delete _pcmi;
    delete _mtdm;
}


void Audiothr::thr_main (void)
{
    int i, k;

    // Start the audio device.
    _pcmi->pcm_start ();

    // Main loop.
    while (!_stop)
    {
	k = _pcmi->pcm_wait ();  
	if (k < _frsize)
	{
	    // Normally we shouldn't do this in a real-time context.
	    fprintf (stderr, "Error: pcm_wait returned %d.", k);
	}
        while (k >= _frsize)
       	{
	    // Read input signal.
            _pcmi->capt_init (_frsize);
            _pcmi->capt_chan (_inp, _ipbuf, _frsize);
            _pcmi->capt_done (_frsize);

	    // Process.
	    _mtdm->process (_frsize, _ipbuf, _opbuf);

	    // Write output signal and clear all other outputs.
            _pcmi->play_init (_frsize);
            _pcmi->play_chan (_out, _opbuf, _frsize);              
	    for (i = 0; i < _pcmi->nplay (); i++)
	    {
		if (i != _out) _pcmi->clear_chan (i, _frsize);
	    }
            _pcmi->play_done (_frsize);

            k -= _frsize;
	}
    }

    // Stop the audio device.
    _pcmi->pcm_stop ();
}


static void sigint_handler (int)
{
    signal (SIGINT, SIG_IGN);
    audio->stop ();
}


int main (int ac, char *av [])
{
    int       fsamp, frsize, nfrags, inp, out;
    MTDM     *mtdm;
    float     t;

    if (ac < 8)
    {
	fprintf (stderr, "alsa-delay <playdev> <captdev> <fsamp> <frsize> <nfrags> <input> <output>\n");
	fprintf (stderr, "Input and output numbers start at 1.\n");
        return 1;
    }

    fsamp = atoi (av [3]);
    frsize = atoi (av [4]);
    nfrags = atoi (av [5]);
    inp = atoi (av [6]) - 1;
    out = atoi (av [7]) - 1;
    
    audio = new Audiothr (av [1], av [2], fsamp, frsize, nfrags, inp, out);
    mtdm = audio->mtdm ();
    t = 1000.0f / fsamp;

    signal (SIGINT, sigint_handler);

    while (audio->running ())
    {
	usleep (250000);

	if (mtdm->resolve () < 0) printf ("Signal below threshold...\n");
	else 
	{
	    if (mtdm->err () > 0.3) 
	    {
		mtdm->invert ();
		mtdm->resolve ();
	    }
	    printf ("%10.3lf frames %10.3lf ms", mtdm->del (), mtdm->del () * t);
	    if (mtdm->err () > 0.2) printf (" ??");
            if (mtdm->inv ()) printf (" Inv");
	    printf ("\n");
	}
    }

    usleep (250000);
    delete audio;
    return 0;
}


// --------------------------------------------------------------------------------



libs/zita-alsa-pcmi.h
// ----------------------------------------------------------------------------
//
//  Copyright (C) 2006-2012 Fons Adriaensen <fons@linuxaudio.org>
//    
//  This program is free software; you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation; either version 3 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
// ----------------------------------------------------------------------------


#ifndef __ZITA_ALSA_PCMI_H
#define __ZITA_ALSA_PCMI_H


#define ALSA_PCM_NEW_HW_PARAMS_API
#define ALSA_PCM_NEW_SW_PARAMS_API
#include <alsa/asoundlib.h>


#define ZITA_ALSA_PCMI_MAJOR_VERSION 0
#define ZITA_ALSA_PCMI_MINOR_VERSION 2


extern int zita_alsa_pcmi_major_version (void);
extern int zita_alsa_pcmi_minor_version (void);


class Alsa_pcmi
{
public:

    Alsa_pcmi (const char        *play_name,
               const char        *capt_name,
   	       const char        *ctrl_name,
               unsigned int       rate,
               unsigned int       frsize,
               unsigned int       nfrags,
	       unsigned int       debug = 0);

    ~Alsa_pcmi (void);  

    enum 
    {
	DEBUG_INIT = 1,
	DEBUG_STAT = 2,
	DEBUG_WAIT = 4,
	DEBUG_DATA = 8,
	DEBUG_ALL  = 15,
	FORCE_16B  = 256,
	FORCE_2CH  = 512
    };

    void printinfo (void);

    int pcm_start (void);
    int pcm_stop (void);
    snd_pcm_sframes_t pcm_wait (void);
    int pcm_idle (int len);

    int play_init (snd_pcm_uframes_t len);
    void clear_chan (int chan, int len);
    void play_chan (int chan, const float *src, int len, int step = 1);
    int play_done (int len);

    int capt_init (snd_pcm_uframes_t len);
    void capt_chan (int chan, float *dst, int len, int step = 1);
    int capt_done (int len);

    int play_avail (void)
    {
	return snd_pcm_avail (_play_handle);
    }

    int capt_avail (void)
    {
	return snd_pcm_avail (_capt_handle);
    }

    int play_delay (void)
    {
	long k;
	snd_pcm_delay (_play_handle, &k);
	return k;
    }

    int capt_delay (void)
    {
	long k;
	snd_pcm_delay (_capt_handle, &k);
	return k;
    }

    float play_xrun (void) const { return _play_xrun; }
    float capt_xrun (void) const { return _capt_xrun; }

    int state (void) const { return _state; }
    int fsamp (void) const { return _fsamp; } 
    int fsize (void) const { return _fsize; } 
    int nfrag (void) const { return _nfrag; } 
    int nplay (void) const { return _play_nchan; }
    int ncapt (void) const { return _capt_nchan; }
    snd_pcm_t *play_handle (void) const { return _play_handle; }
    snd_pcm_t *capt_handle (void) const { return _capt_handle; }
    

private:

    typedef char *(Alsa_pcmi::*clear_function)(char *, int);
    typedef char *(Alsa_pcmi::*play_function)(const float *, char *, int, int);
    typedef const char *(Alsa_pcmi::*capt_function) (const char *, float *, int, int);

    enum { MAXPFD = 16, MAXCHAN = 64 };

    void initialise (const char *play_name, const char *capt_name, const char *ctrl_name);
    int set_hwpar (snd_pcm_t *handle, snd_pcm_hw_params_t *hwpar, const char *sname, unsigned int *nchan);
    int set_swpar (snd_pcm_t *handle, snd_pcm_sw_params_t *swpar, const char *sname);
    int recover (void);
    float xruncheck (snd_pcm_status_t *stat);

    char *clear_32 (char *dst, int nfrm);
    char *clear_24 (char *dst, int nfrm);
    char *clear_16 (char *dst, int nfrm);

    char *play_float (const float *src, char *dst, int nfrm, int step);
    char *play_32 (const float *src, char *dst, int nfrm, int step);
    char *play_24 (const float *src, char *dst, int nfrm, int step);
    char *play_16 (const float *src, char *dst, int nfrm, int step);
    char *play_32swap (const float *src, char *dst, int nfrm, int step);
    char *play_24swap (const float *src, char *dst, int nfrm, int step);
    char *play_16swap (const float *src, char *dst, int nfrm, int step);

    const char *capt_float (const char *src, float *dst, int nfrm, int step);
    const char *capt_32 (const char *src, float *dst, int nfrm, int step);
    const char *capt_24 (const char *src, float *dst, int nfrm, int step);
    const char *capt_16 (const char *src, float *dst, int nfrm, int step);
    const char *capt_32swap (const char *src, float *dst, int nfrm, int step);
    const char *capt_24swap (const char *src, float *dst, int nfrm, int step);
    const char *capt_16swap (const char *src, float *dst, int nfrm, int step);

    unsigned int           _fsamp;
    snd_pcm_uframes_t      _fsize;
    unsigned int           _nfrag;
    unsigned int           _debug;
    int                    _state;
    snd_pcm_t             *_play_handle;
    snd_pcm_t             *_capt_handle;
    snd_ctl_t             *_ctrl_handle;
    snd_pcm_hw_params_t   *_play_hwpar;
    snd_pcm_sw_params_t   *_play_swpar;
    snd_pcm_hw_params_t   *_capt_hwpar;
    snd_pcm_sw_params_t   *_capt_swpar;
    snd_pcm_format_t       _play_format;
    snd_pcm_format_t       _capt_format;
    snd_pcm_access_t       _play_access;
    snd_pcm_access_t       _capt_access;
    unsigned int           _play_nchan;
    unsigned int           _capt_nchan;
    float                  _play_xrun;
    float                  _capt_xrun;
    bool                   _synced;
    int                    _play_npfd;
    int                    _capt_npfd;
    struct pollfd          _poll_fd [MAXPFD];
    snd_pcm_uframes_t      _capt_offs;
    snd_pcm_uframes_t      _play_offs;
    int                    _play_step;
    int                    _capt_step;
    char                  *_play_ptr [MAXCHAN];
    const char            *_capt_ptr [MAXCHAN];
    clear_function         _clear_func;
    play_function          _play_func;
    capt_function          _capt_func;
    void                  *_dummy [16];
};


#endif



apps/alsa_loopback.cc
// ----------------------------------------------------------------------------
//
//  Copyright (C) 2006-2018 Fons Adriaensen <fons@linuxaudio.org>
//    
//  This program is free software; you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation; either version 3 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
// ----------------------------------------------------------------------------


#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <zita-alsa-pcmi.h>
#include "pxthread.h"


// --------------------------------------------------------------------------------


static char  playdev [256];
static char  captdev [256];
static int   fsamp;
static int   frsize;
static int   nfrags;


class Audiothr : public Pxthread
{
private:
 
    virtual void thr_main (void);
};


void Audiothr::thr_main (void)
{
    Alsa_pcmi  *D;
    float      *buf;
    int        i, k;

    // Interleaved buffer for channels 1 and 2.
    buf = new float [frsize * 2];

    // Create and initialise the audio device.
    D = new Alsa_pcmi (playdev, captdev, 0, fsamp, frsize, nfrags, 0);
    if (D->state ()) 
    {
	fprintf (stderr, "Can't open ALSA device\n");
        delete D;
        exit (1);
    }
    if ((D->ncapt () < 2) || (D->nplay () < 2))
    {
	fprintf (stderr, "Expected a stereo device.\n");
        delete D;
        exit (1);
    }
    D->printinfo ();

    // Start the audio device.
    D->pcm_start ();

    // Main loop
    while (true)
    {
	k = D->pcm_wait ();  
	if (k < frsize)
	{
	    // Normally we shouldn't do this in a real-time context.
	    fprintf (stderr, "Error: pcm_wait returned %d.\n", k);
	}
        while (k >= frsize)
       	{
	    // Copy the first 2 inputs to the first two outputs.
	    // Clear all other outputs.
            D->capt_init (frsize);
            D->capt_chan (0, buf + 0, frsize, 2);
            D->capt_chan (1, buf + 1, frsize, 2);              
            D->capt_done (frsize);

            D->play_init (frsize);
            D->play_chan (0, buf + 0, frsize, 2);
            D->play_chan (1, buf + 1, frsize, 2);              
	    for (i = 2; i < D->nplay (); i++) D->clear_chan (i, frsize);
            D->play_done (frsize);

            k -= frsize;
	}
    }

    // Stop the audio device.
    D->pcm_stop ();

    delete D;
    delete[] buf;
}


// --------------------------------------------------------------------------------


int main (int ac, char *av [])
{
    Audiothr A;

    if (ac < 6)
    {
	fprintf (stderr, "alsa-loopback <playdev> <captdev> <fsamp> <frsize> <nfrags>\n");
        return 1;
    }

    strcpy (playdev, av [1]);
    strcpy (captdev, av [2]);
    fsamp  = atoi (av [3]);
    frsize = atoi (av [4]);
    nfrags = atoi (av [5]);

    // Run a real-time thread to handle audio.
    if (A.thr_start (SCHED_FIFO, -50, 0x20000))
    {
 	fprintf (stderr, "Can't run in RT mode, trying normal scheduling.\n");
        if (A.thr_start (SCHED_OTHER, 0, 0x20000))
        {
 	    fprintf (stderr, "Can't start audio thread.\n");
	    return 1;
	}
    }

    sleep (1000000);

    return 0;
}


// --------------------------------------------------------------------------------
